From 7cccd00e4d967518f1e2ddfc54b988836e34ab0d Mon Sep 17 00:00:00 2001
From: Jerome Martinez <jerome@mediaarea.net>
Date: Fri, 30 Oct 2026 07:50:56 +0100
Subject: [PATCH] avcodec/tiff: support 6, 10 and 14-bit gray, 2, 4, 6, 10, 12
 and 14-bit RGB

Gray with 6, 10 or 14 bits per sample is decoded to GRAY8, GRAY10 or
GRAY14, RGB with 2, 4, 6, 10, 12 or 14 bits per sample to GBRP, GBRP10,
GBRP12 or GBRP14, also with separate planes. Samples with less than 8
bits are scaled to 8 bits like the gray palette.
12-bit gray lines are now also unpacked with deflate and LZMA
compression, they were decompressed into the line buffer but never
unpacked into the frame.
---
 libavcodec/tiff.c | 109 +++++++++++++++++++++++++++++++++++++++-------
 1 file changed, 94 insertions(+), 15 deletions(-)

diff --git a/libavcodec/tiff.c b/libavcodec/tiff.c
index efe7ae7056..314d29df55 100644
--- a/libavcodec/tiff.c
+++ b/libavcodec/tiff.c
@@ -455,17 +455,59 @@ static int deinvert_buffer(TiffContext *s, const uint8_t *src, int size)
     return 0;
 }
 
-static void unpack_gray(TiffContext *s, AVFrame *p,
-                       const uint8_t *src, int lnum, int width, int bpp)
+/* Samples with a bit depth other than 8 and 16, unpacked from a line buffer */
+static int is_unpacked(const TiffContext *s, const AVFrame *p)
+{
+    switch (p->format) {
+    case AV_PIX_FMT_GRAY10:
+    case AV_PIX_FMT_GRAY12:
+    case AV_PIX_FMT_GRAY14:
+    case AV_PIX_FMT_GBRP10:
+    case AV_PIX_FMT_GBRP12:
+    case AV_PIX_FMT_GBRP14:
+        return 1;
+    case AV_PIX_FMT_GRAY8:
+    case AV_PIX_FMT_GBRP:
+        return s->bpp / s->bppcount < 8;
+    default:
+        return 0;
+    }
+}
+
+/* Samples with less than 8 bits are scaled to 8 bits like the gray palette */
+static void unpack_line(TiffContext *s, uint8_t *const *dst, int components,
+                        const uint8_t *src, int width)
 {
     GetBitContext gb;
-    uint16_t *dst = (uint16_t *)(p->data[0] + lnum * p->linesize[0]);
+    int bps = s->bpp / s->bppcount;
 
     av_unused int ret = init_get_bits8(&gb, src, width);
     av_assert1(ret >= 0);
 
     for (int i = 0; i < s->width; i++) {
-        dst[i] = get_bits(&gb, bpp);
+        for (int c = 0; c < components; c++) {
+            unsigned v = get_bits(&gb, bps);
+            if (bps > 8)
+                ((uint16_t *)dst[c])[i] = v;
+            else
+                dst[c][i] = v * 255 / ((1 << bps) - 1);
+        }
+    }
+}
+
+/* out: line of the plane for planar or gray, lnum: line of the frame for RGB */
+static void unpack(TiffContext *s, AVFrame *p, uint8_t *out,
+                   const uint8_t *src, int lnum, int width)
+{
+    if (s->planar || s->bppcount == 1) {
+        unpack_line(s, &out, 1, src, width);
+    } else {
+        uint8_t *rgb[3] = {
+            p->data[2] + lnum * p->linesize[2],
+            p->data[0] + lnum * p->linesize[0],
+            p->data[1] + lnum * p->linesize[1],
+        };
+        unpack_line(s, rgb, 3, src, width);
     }
 }
 
@@ -520,6 +562,7 @@ static int tiff_uncompress(uint8_t *dst, unsigned long *len, const uint8_t *src,
 }
 
 static int tiff_unpack_zlib(TiffContext *s, AVFrame *p, uint8_t *dst, int stride,
+                            uint8_t *out, int out_stride,
                             const uint8_t *src, int size, int width, int lines,
                             int strip_start, int is_yuv)
 {
@@ -562,6 +605,8 @@ static int tiff_unpack_zlib(TiffContext *s, AVFrame *p, uint8_t *dst, int stride
         if (is_yuv) {
             unpack_yuv(s, p, dst, strip_start + line);
             line += s->subsampling[1] - 1;
+        } else if (is_unpacked(s, p)) {
+            unpack(s, p, out + line * out_stride, dst, strip_start + line, width);
         }
         dst += stride;
         src += width;
@@ -594,6 +639,7 @@ static int tiff_uncompress_lzma(uint8_t *dst, uint64_t *len, const uint8_t *src,
 }
 
 static int tiff_unpack_lzma(TiffContext *s, AVFrame *p, uint8_t *dst, int stride,
+                            uint8_t *out, int out_stride,
                             const uint8_t *src, int size, int width, int lines,
                             int strip_start, int is_yuv)
 {
@@ -634,6 +680,8 @@ static int tiff_unpack_lzma(TiffContext *s, AVFrame *p, uint8_t *dst, int stride
         if (is_yuv) {
             unpack_yuv(s, p, dst, strip_start + line);
             line += s->subsampling[1] - 1;
+        } else if (is_unpacked(s, p)) {
+            unpack(s, p, out + line * out_stride, dst, strip_start + line, width);
         }
         dst += stride;
         src += width;
@@ -761,6 +809,8 @@ static int tiff_unpack_strip(TiffContext *s, AVFrame *p, uint8_t *dst, int strid
     int c, line, pixels, code, ret;
     const uint8_t *ssrc = src;
     int width = ((s->width * s->bpp) + 7) >> 3;
+    uint8_t *out = dst;
+    int out_stride = stride;
     const AVPixFmtDescriptor *desc = av_pix_fmt_desc_get(p->format);
     int is_yuv = !(desc->flags & AV_PIX_FMT_FLAG_RGB) &&
                  (desc->flags & AV_PIX_FMT_FLAG_PLANAR) &&
@@ -768,7 +818,7 @@ static int tiff_unpack_strip(TiffContext *s, AVFrame *p, uint8_t *dst, int strid
     int is_dng;
 
     if (s->planar)
-        width /= s->bppcount;
+        width = ((s->width * (s->bpp / s->bppcount)) + 7) >> 3;
 
     if (size <= 0)
         return AVERROR_INVALIDDATA;
@@ -793,7 +843,7 @@ static int tiff_unpack_strip(TiffContext *s, AVFrame *p, uint8_t *dst, int strid
         av_assert0(width == (s->bpp * s->width + 7) >> 3);
     }
     av_assert0(!(s->is_bayer && is_yuv));
-    if (p->format == AV_PIX_FMT_GRAY12) {
+    if (is_unpacked(s, p)) {
         av_fast_padded_malloc(&s->yuv_line, &s->yuv_line_size, width);
         if (s->yuv_line == NULL) {
             av_log(s->avctx, AV_LOG_ERROR, "Not enough memory\n");
@@ -805,8 +855,8 @@ static int tiff_unpack_strip(TiffContext *s, AVFrame *p, uint8_t *dst, int strid
 
     if (s->compr == TIFF_DEFLATE || s->compr == TIFF_ADOBE_DEFLATE) {
 #if CONFIG_ZLIB
-        return tiff_unpack_zlib(s, p, dst, stride, src, size, width, lines,
-                                strip_start, is_yuv);
+        return tiff_unpack_zlib(s, p, dst, stride, out, out_stride, src, size,
+                                width, lines, strip_start, is_yuv);
 #else
         av_log(s->avctx, AV_LOG_ERROR,
                "zlib support not enabled, "
@@ -816,8 +866,8 @@ static int tiff_unpack_strip(TiffContext *s, AVFrame *p, uint8_t *dst, int strid
     }
     if (s->compr == TIFF_LZMA) {
 #if CONFIG_LZMA
-        return tiff_unpack_lzma(s, p, dst, stride, src, size, width, lines,
-                                strip_start, is_yuv);
+        return tiff_unpack_lzma(s, p, dst, stride, out, out_stride, src, size,
+                                width, lines, strip_start, is_yuv);
 #else
         av_log(s->avctx, AV_LOG_ERROR,
                "LZMA support not enabled\n");
@@ -849,8 +899,8 @@ static int tiff_unpack_strip(TiffContext *s, AVFrame *p, uint8_t *dst, int strid
             if (is_yuv) {
                 unpack_yuv(s, p, dst, strip_start + line);
                 line += s->subsampling[1] - 1;
-            } else if (p->format == AV_PIX_FMT_GRAY12) {
-                unpack_gray(s, p, dst, strip_start + line, width, s->bpp);
+            } else if (is_unpacked(s, p)) {
+                unpack(s, p, out + line * out_stride, dst, strip_start + line, width);
             }
             dst += stride;
         }
@@ -859,7 +909,7 @@ static int tiff_unpack_strip(TiffContext *s, AVFrame *p, uint8_t *dst, int strid
     if (s->compr == TIFF_CCITT_RLE ||
         s->compr == TIFF_G3        ||
         s->compr == TIFF_G4) {
-        if (is_yuv || p->format == AV_PIX_FMT_GRAY12)
+        if (is_yuv || is_unpacked(s, p))
             return AVERROR_INVALIDDATA;
 
         return tiff_unpack_fax(s, dst, stride, src, size, width, lines);
@@ -975,8 +1025,8 @@ static int tiff_unpack_strip(TiffContext *s, AVFrame *p, uint8_t *dst, int strid
         if (is_yuv) {
             unpack_yuv(s, p, dst, strip_start + line);
             line += s->subsampling[1] - 1;
-        } else if (p->format == AV_PIX_FMT_GRAY12) {
-            unpack_gray(s, p, dst, strip_start + line, width, s->bpp);
+        } else if (is_unpacked(s, p)) {
+            unpack(s, p, out + line * out_stride, dst, strip_start + line, width);
         }
         dst += stride;
     }
@@ -1083,12 +1133,29 @@ static int init_image(TiffContext *s, AVFrame *frame)
             create_gray_palette = 1;
         }
         break;
+    case 61:
+        s->avctx->pix_fmt = AV_PIX_FMT_GRAY8;
+        break;
+    case 63:
+    case 123:
+    case 183:
+    case 10063:
+    case 10123:
+    case 10183:
+        s->avctx->pix_fmt = AV_PIX_FMT_GBRP;
+        break;
     case 81:
         s->avctx->pix_fmt = s->palette_is_set ? AV_PIX_FMT_PAL8 : AV_PIX_FMT_GRAY8;
         break;
+    case 101:
+        s->avctx->pix_fmt = AV_PIX_FMT_GRAY10;
+        break;
     case 121:
         s->avctx->pix_fmt = AV_PIX_FMT_GRAY12;
         break;
+    case 141:
+        s->avctx->pix_fmt = AV_PIX_FMT_GRAY14;
+        break;
     case 100081:
         switch (AV_RL32(s->pattern)) {
         case 0x02010100:
@@ -1165,6 +1232,18 @@ static int init_image(TiffContext *s, AVFrame *frame)
     case 324:
         s->avctx->pix_fmt = s->photometric == TIFF_PHOTOMETRIC_SEPARATED ? AV_PIX_FMT_RGB0 : AV_PIX_FMT_RGBA;
         break;
+    case 303:
+    case 10303:
+        s->avctx->pix_fmt = AV_PIX_FMT_GBRP10;
+        break;
+    case 363:
+    case 10363:
+        s->avctx->pix_fmt = AV_PIX_FMT_GBRP12;
+        break;
+    case 423:
+    case 10423:
+        s->avctx->pix_fmt = AV_PIX_FMT_GBRP14;
+        break;
     case 405:
         if (s->photometric == TIFF_PHOTOMETRIC_SEPARATED)
             s->avctx->pix_fmt = AV_PIX_FMT_RGBA;
-- 
2.46.0.windows.1

